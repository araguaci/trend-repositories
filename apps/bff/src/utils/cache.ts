const LRU = require('lru-cache');

export let cache = null;

export const initCache = () => {
  const options = {
    max: 500,
    ttl: 86400 * 1000, // 1 day
  };

  cache = new LRU(options);
};
