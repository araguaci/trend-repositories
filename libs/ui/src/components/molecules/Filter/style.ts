import styled from 'styled-components';

export const StyledFilter = styled.div`
  display: flex;
  justify-content: space-between;
  padding: ${(props) => props.theme.sizes.medium};
  background-color: ${(props) => props.theme.colors.black_2};
  border-top-left-radius: 6px;
  border-top-right-radius: 6px;

  @media only screen and (max-width: 600px) {
    flex-direction: column;
    align-items: center;

    > div:first-child {
      margin-bottom: 9px;
    }
  }
`;

export const ButtonContainer = styled.div`
  a:nth-child(1) {
    border-bottom-right-radius: 0;
    border-top-right-radius: 0;
    border-right: none;
  }

  a:nth-child(2) {
    border-bottom-left-radius: 0;
    border-top-left-radius: 0;
    border-left: none;
  }

  a:hover {
    text-decoration: none;
  }
`;

export const SelectContainer = styled.div`
  display: flex;
  align-items: center;

  select {
    background-color: ${(props) => props.theme.colors.grey_2};
    color: ${(props) => props.theme.colors.white_2};
    border-radius: 6px;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    max-width: 100px;
    margin-right: 10px;
  }
`;
