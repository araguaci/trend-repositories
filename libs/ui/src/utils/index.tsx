/* istanbul ignore file */
import { ThemeProvider } from 'styled-components';
import theme from '../theme';
import { ReactNode } from 'react';
import { render, act } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { MockedResponse } from '@apollo/client/testing/core';
import { InMemoryCache } from '@apollo/client/cache';

export const renderWithProvider = (node: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>);
};

export const renderWithMockProvider = (
  node: ReactNode,
  cache: InMemoryCache,
  mocks: ReadonlyArray<MockedResponse>
) => {
  return render(
    <ThemeProvider theme={theme}>
      <MockedProvider cache={cache} mocks={mocks}>
        {node}
      </MockedProvider>
    </ThemeProvider>
  );
};

export const renderWithThemeProvider = (node: ReactNode) => {
  return render(<ThemeProvider theme={theme}>{node}</ThemeProvider>);
};

export const waitForResponse = async () =>
  await act(async () => await new Promise((resolve) => setTimeout(resolve, 0)));
export * from '@testing-library/react';
