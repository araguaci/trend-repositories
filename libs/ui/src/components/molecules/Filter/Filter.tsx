import { Button } from '../../atoms';
import { ButtonContainer, SelectContainer, StyledFilter } from './style';
import { FC, SyntheticEvent } from 'react';
import { useRouter } from 'next/router';
import { Language } from '@trend-repositories/api-interfaces';

interface FilterProps {
  languages: Language[];
}

const Filter: FC<FilterProps> = ({ languages }) => {
  const router = useRouter();
  const { filter, language, date } = router.query;

  const handleLanguageChange = (event: SyntheticEvent<HTMLSelectElement>) => {
    router.push({
      pathname: '/',
      query: { ...router.query, language: event.currentTarget.value, page: 1 },
    });
  };

  const handleDateChange = (event: SyntheticEvent<HTMLSelectElement>) => {
    router.push({
      pathname: '/',
      query: { ...router.query, date: event.currentTarget.value, page: 1 },
    });
  };

  return (
    <StyledFilter>
      <ButtonContainer>
        <Button
          active={filter !== 'star'}
          href={{
            pathname: '/',
            query: { ...router.query, filter: 'all' },
          }}
        >
          Repositories
        </Button>
        <Button
          data-cy="filter-star"
          active={filter === 'star'}
          href={{
            pathname: '/',
            query: { ...router.query, filter: 'star' },
          }}
        >
          Favorite Repositories
        </Button>
      </ButtonContainer>

      <SelectContainer>
        <select
          data-cy="language-selection"
          aria-label="language-selection"
          onChange={handleLanguageChange}
          value={language}
        >
          <option value="all">All Language</option>
          {languages.map((lang) => (
            <option key={lang.text} value={lang.text}>
              {lang.text}
            </option>
          ))}
        </select>

        <select
          data-cy="date-selection"
          aria-label="date-selection"
          onChange={handleDateChange}
          value={date}
        >
          <option hidden>Date Range</option>
          <option value="today">Today</option>
          <option value="last_week">Last week</option>
          <option value="last_month">Last month</option>
        </select>
      </SelectContainer>
    </StyledFilter>
  );
};

export default Filter;
