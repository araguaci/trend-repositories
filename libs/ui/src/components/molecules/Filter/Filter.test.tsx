import singletonRouter from 'next/router';
import mockRouter from 'next-router-mock';
import { renderWithThemeProvider, screen, fireEvent } from '@ui/utils';
import Filter from './Filter';

jest.mock('next/dist/client/router', () => require('next-router-mock'));

const languages = [
  { color: 'yellow', text: 'Javascript' },
  { color: 'red', text: 'PHP' },
];

describe('Filter Component Test', () => {
  beforeEach(() => {
    mockRouter.setCurrentUrl('/');
  });

  it('should be rendered correctly', () => {
    // Arrange && Act
    const { asFragment } = renderWithThemeProvider(
      <Filter languages={languages} />
    );

    // Assertion
    expect(asFragment()).toMatchSnapshot();
  });

  it('should filter all repositories', () => {
    // Arrange
    renderWithThemeProvider(<Filter languages={languages} />);

    const repositoriesButton = screen.getByRole('link', {
      name: 'Repositories',
    });

    // Act
    fireEvent.click(repositoriesButton);

    // Assertion
    expect(singletonRouter).toMatchObject({
      asPath: '/?filter=all',
      pathname: '/',
      query: { filter: 'all' },
    });
  });

  it('should filter starred repositories', () => {
    // Arrange
    renderWithThemeProvider(<Filter languages={languages} />);

    const repositoriesButton = screen.getByRole('link', {
      name: 'Favorite Repositories',
    });

    // Act
    fireEvent.click(repositoriesButton);

    // Assertion
    expect(singletonRouter).toMatchObject({
      asPath: '/?filter=star',
      pathname: '/',
      query: { filter: 'star' },
    });
  });

  it('should filter by language', () => {
    // Arrange
    renderWithThemeProvider(<Filter languages={languages} />);

    const languageCombo = screen.getByRole('combobox', {
      name: 'language-selection',
    });

    // Act
    fireEvent.change(languageCombo, { target: { value: 'PHP' } });

    // Assertion
    expect(singletonRouter).toMatchObject({
      asPath: '/?language=PHP&page=1',
      pathname: '/',
      query: { language: 'PHP', page: 1 },
    });
  });

  it('should filter by date', () => {
    // Arrange
    renderWithThemeProvider(<Filter languages={languages} />);

    const languageCombo = screen.getByRole('combobox', {
      name: 'date-selection',
    });

    // Act
    fireEvent.change(languageCombo, { target: { value: 'last_month' } });

    // Assertion
    expect(singletonRouter).toMatchObject({
      asPath: '/?date=last_month&page=1',
      pathname: '/',
      query: { page: 1 },
    });
  });
});
