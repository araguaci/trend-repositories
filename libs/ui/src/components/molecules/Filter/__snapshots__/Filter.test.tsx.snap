// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Component Test should be rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="style__StyledFilter-sc-1pyayj-0 fQCbPz"
  >
    <div
      class="style__ButtonContainer-sc-1pyayj-1 iqZZMY"
    >
      <a
        class="style__StyledLink-ikww7b-0 fUma-dV style__StyledButton-gn6n84-0 fyOAYP"
        href="/?filter=all"
      >
        Repositories
      </a>
      <a
        class="style__StyledLink-ikww7b-0 fUma-dV style__StyledButton-gn6n84-0 eAVKfN"
        data-cy="filter-star"
        href="/?filter=star"
      >
        Favorite Repositories
      </a>
    </div>
    <div
      class="style__SelectContainer-sc-1pyayj-2 clhOTY"
    >
      <select
        aria-label="language-selection"
        data-cy="language-selection"
      >
        <option
          value="all"
        >
          All Language
        </option>
        <option
          value="Javascript"
        >
          Javascript
        </option>
        <option
          value="PHP"
        >
          PHP
        </option>
      </select>
      <select
        aria-label="date-selection"
        data-cy="date-selection"
      >
        <option
          hidden=""
        >
          Date Range
        </option>
        <option
          value="today"
        >
          Today
        </option>
        <option
          value="last_week"
        >
          Last week
        </option>
        <option
          value="last_month"
        >
          Last month
        </option>
      </select>
    </div>
  </div>
</DocumentFragment>
`;
